### ISSUE TITLE
NotImplementedError: unknown node: autosummary_table

### ISSUE THREAD

**Issue by engineerjoe440** [sphinx-contrib/confluencebuilder#304]
Ran into an issue upon my first test-run with this module. From other issues, it seems like this has already been managed, so maybe I'm just missing something. I'm still pretty sphinx illiterate, so any help would be greatly appreciated!

Full Traceback:

```
# Sphinx version: 2.3.1
# Python version: 3.8.1 (CPython)
# Docutils version: 0.16 release
# Jinja2 version: 2.10.3
# Last messages:
#   
#   looking for now-outdated files...
#   none found
#   pickling environment...
#   done
#   checking consistency...
#   done
#   preparing documents...
#   done
#   writing output... [ 14%] docs/AcRTAC
# Loaded extensions:
#   sphinx.ext.mathjax (2.3.1) from c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinx\ext\mathjax.py
#   sphinxcontrib.applehelp (1.0.1) from c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinxcontrib\applehelp\__init__.py
#   sphinxcontrib.devhelp (1.0.1) from c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinxcontrib\devhelp\__init__.py
#   sphinxcontrib.htmlhelp (1.0.2) from c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinxcontrib\htmlhelp\__init__.py
#   sphinxcontrib.serializinghtml (1.1.3) from c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinxcontrib\serializinghtml\__init__.py
#   sphinxcontrib.qthelp (1.0.2) from c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinxcontrib\qthelp\__init__.py
#   alabaster (0.7.12) from c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\alabaster\__init__.py
#   sphinx.ext.autodoc (2.3.1) from c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinx\ext\autodoc\__init__.py
#   sphinx.ext.napoleon (2.3.1) from c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinx\ext\napoleon\__init__.py
#   sphinx.ext.autosummary (2.3.1) from c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinx\ext\autosummary\__init__.py
#   numpydoc (0.9.2) from c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\numpydoc\__init__.py
#   sphinxcontrib.confluencebuilder (1.2.0) from c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinxcontrib\confluencebuilder\__init__.py
Traceback (most recent call last):
  File "c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinx\cmd\build.py", line 276, in build_main
    app.build(args.force_all, filenames)
  File "c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinx\application.py", line 353, in build
    self.builder.build_update()
  File "c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinx\builders\__init__.py", line 297, in build_update
    self.build(to_build,
  File "c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinx\builders\__init__.py", line 361, in build
    self.write(docnames, list(updated_docnames), method)
  File "c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinx\builders\__init__.py", line 535, in write
    self._write_serial(sorted(docnames))
  File "c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinx\builders\__init__.py", line 545, in _write_serial
    self.write_doc(docname, doctree)
  File "c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinxcontrib\confluencebuilder\builder.py", line 435, in write_doc
    self.writer.write(doctree, destination)
  File "c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\docutils\writers\__init__.py", line 78, in write
    self.translate()
  File "c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinxcontrib\confluencebuilder\writer.py", line 24, in translate
    self.document.walkabout(visitor)
  File "c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\docutils\nodes.py", line 214, in walkabout
    if child.walkabout(visitor):
  File "c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\docutils\nodes.py", line 214, in walkabout
    if child.walkabout(visitor):
  File "c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\docutils\nodes.py", line 214, in walkabout
    if child.walkabout(visitor):
  [Previous line repeated 2 more times]
  File "c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\docutils\nodes.py", line 206, in walkabout
    visitor.dispatch_visit(self)
  File "c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\docutils\nodes.py", line 1995, in dispatch_visit
    return method(node)
  File "c:\users\joestan\appdata\local\programs\python\python38-32\lib\site-packages\sphinxcontrib\confluencebuilder\translator.py", line 149, in unknown_visit
    raise NotImplementedError('unknown node: ' + node_name)
NotImplementedError: unknown node: autosummary_table
```

For reference, here's a text-file copy of the AcRTAC.rst file that seems to be the source of my trouble.

[AcRTAC.rst.txt](https://github.com/sphinx-contrib/confluencebuilder/files/4257540/AcRTAC.rst.txt)

**Comment by jdknight** [comment]
Interesting. Support for [`sphinx.ext.autosummary`](https://www.sphinx-doc.org/en/master/usage/extensions/autosummary.html#directive-autosummary) capabilities _should_ be working, but clearly there is a use case where there is an issue (thanks for reporting it). The call `autosummary_table` should be supported in the builder as this extension [injects](https://github.com/sphinx-contrib/confluencebuilder/blob/v1.2/sphinxcontrib/confluencebuilder/__init__.py#L201-L202) autosummary's node handlers into the translator during initialization.

From the provided output, it does appear that `sphinx.ext.autosummary` is a loaded extension. The only thing I can think of off hand is maybe `sphinx.ext.autosummary` is loaded implicitly and the check for the extension inside `app.config.extensions` will always fail since it is not explicitly in the list (speculation; implicitly loaded extension may not even be a concept in Sphinx). A workaround for now, assuming the environment do not have `sphinx.ext.autosummary` in the extension list, is to try to add it and see if you get different results. In the rare case that this is it, this extension's implementation should be probably tweaked how it checks for loaded extensions.

In the more likely case the above does not work, I would not have a solution to resolve this at this time. I will try to find some time to test with the example document to reproduce the issue locally.

A workaround to skip the autosummary nodes for an interim could be performed with the following:

```
confluence_adv_ignore_nodes = ['autosummary_table', 'autosummary_toc']
```

**Comment by engineerjoe440** [comment]
@jdknight, thank you so much for your prompt response. I wish I had responded as quickly. Been so terribly busy I haven't had a chance to circle back around. I'll try to use the workaround you've suggested, and I'll report back!

Thank you again!

**Comment by engineerjoe440** [comment]
@jdknight, that workaround did in fact work perfectly! Thank you so much for your help!!!

**Comment by jannismain** [comment]
I've run into the same issue and solved it by adding `sphinx.ext.autosummary` to my list of `extensions` in `conf.py`, as suggested by @jdknight

**Comment by adrianohrl** [comment]
I've tried to use the `sphinx.ext.graphviz` along with the `sphinxcontrib.confluencebuilder`; although, a `NotImplementedError` is prompted to me.

**Comment by jdknight** [comment]
After some further investigation, finally figured out what would cause the following exception:

```
NotImplementedError: unknown node: autosummary_table
```

For v1.2 of this extension, when `sphinx.ext.autosummary` is loaded into the explicit extension list (in the project's configuration file), it would know that autosummary support was enabled and would map supported autosummary nodes into the extension. This should not have been an issue for cases not using autosummary since these nodes would never attempt to be visited. The situation outlined in the initial issue comment reveals an issue -- somehow this extension could not load map autosummary nodes but the `sphinx.ext.autosummary` is logged to have been loaded:

```
# Loaded extensions:
...
#   sphinx.ext.autosummary (2.3.1)...
```

The problem resides in the Confluence builder extension on how it verifies how extensions are loaded. What was happening above is `numpydoc` extension would implicitly load `sphinx.ext.autosummary`, which in turn would enable autosummary nodes for autodoc-generated content. However, since this Confluence builder extension only checks the configuration's explicit `extensions` key, it did not believe the extension was loaded. This would then cause an "unknown node: autosummary_table" error when any autodocs-related content was processed.

The proper workaround (i.e. not ignoring the nodes) would be to explicitly load the `sphinx.ext.autosummary` extension in the project's `conf.py` file (also mentioned in some of the comments above). A local fix is in place and will be merged into main branch soon, and in turn, the next release.

**Comment by engineerjoe440** [comment]
@jdknight, thanks for digging into this one so hard. That totally makes sense, thank you for all your hard work!!!!

**Comment by jdknight** [comment]
The changes outlined in #370 are now available on the main branch and will be available next stable release (v1.3+). Will mark as pending-next-release and will close up when the release has been completed (see #362 for more details).

**Comment by jdknight** [comment]
v1.3 is now available on PyPI -- marking as closed.
